kitsune/community/utils.py

```diff
@@ -4,56 +4,86 @@
 
-    if start is None:
-        # By default we go back 90 days.
-        start = date.today() - timedelta(days=90)
-        answers = answers.filter(created__gte=start)
-    if end:
-        # If no end is specified, we don't need to filter by it.
-        answers = answers.filter(created__lt=end)
+    filters = [
+        # filter out answers by the question author
+        {"script": {"script": "doc['creator_id'].value != doc['question_creator_id'].value"}},
+        # filter answers created between `start` and `end`, or within the last 90 days
+        {"range": {"created": {"gte": start or datetime.now() - timedelta(days=90), "lte": end}}},
+    ]
     if locale:
-        answers = answers.filter(question__locale=locale)
+        filters.append({"term": {"locale": locale}})
     if product:
-        if isinstance(product, Product):
-            product = product.slug
-        answers = answers.filter(question__product__slug=product)
-
-    users = (
-        User.objects.filter(answers__in=answers)
-        .annotate(query_count=Count("answers"))
-        .order_by("-query_count")
+        filters.append({"term": {"question_product_id": product.id}})
+
+    search.aggs.bucket(
+        # create a single bucket of documents which match the previously defined filters
+        "filters",
+        A("filter", bool={"filter": filters}),
+    ).bucket(
+        # within that, create buckets for the `count + 1` most active contributors
+        "contributions",
+        A("terms", field="creator_id", size=count + 1),
+    ).bucket(
+        # within each of those, create a bucket for the most recent answer, and extract its date
+        "latest",
+        A(
+            "top_hits",
+            sort={"created": {"order": "desc"}},
+            _source={"includes": "created"},
+            size=1,
+        ),
     )
-    counts = _get_creator_counts(users, count, page)
 
-    if use_cache:
-        cache.set(cache_key, counts, 60 * 180)  # 3 hours
-    return counts
+    contribution_buckets = search.execute().aggregations.filters.contributions.buckets
+    total_contributors = len(contribution_buckets)
+
+    user_ids = map(lambda x: int(x.key), contribution_buckets[:count])
+    users = {
+        # fetch the users found in the ES aggregation,
+        # and put them in a dict with their id as the key
+        user["id"]: user
+        for user in User.objects.filter(id__in=user_ids).values(
+            "id", "username", "profile__name", "profile__fxa_avatar"
+        )
+    }
+
+    return (
+        [
+            {
+                "count": bucket.doc_count,
+                "term": bucket.key,
+                "user": {
+                    "id": user["id"],
+                    "username": user["username"],
+                    "display_name": user["profile__name"],
+                    "avatar": user["profile__fxa_avatar"],
+                    "days_since_last_activity": (
+                        datetime.now()
```

Comment:
nit: maybe we calculate this outside of the return statement. My only reason for proposing this is code readability